python/ray/experimental/sgd/sgd.md

```diff
@@ -0,0 +1,401 @@
+
+1. Hyperparameter tuning -- you want to split the dataset differently per trial.
+2. There's no forced ordering for dataset / trainer creation. (otherwise, you maybe to design your app to create data after instantiating your trainer, and maybe it's not so easy to do that.)
+
+```python
+
+    
+# Preferred API
+def train_func(config):
+    batch_size = config["batch_size"]
+    if ray.sgd.in_worker():
+        dataloader = ray.sgd.get_dataset().get_split().to_torch(batch_size=batch_size)
+    else:
+        dataloader = get_dataset()
+    ...
+    return model
+
+trainer = Trainer(num_workers=8, backend="torch")
+dataset = ray.dataset.filter().pipeline(length=50) # do not call split!
+trainer.run(train_func, config={}, dataset=dataset)
+
+```
+
+Cross Validation:
+
+```python
+
+cv_splits = ray.dataset.to_cv(splits=5)
+
+for train_set, val_set in cv_splits:
+    result = trainer.run(train_func, dataset=train_set)
+    result = trainer.run(train_func, dataset=val_set)
+```
+
+<details>
+<summary> Alternative API </summary>
+
+```python
+# This was not chosen because it requires the user to think about 
+# too many points - both the training function signature and the Trainer declaration.
+
+# def train_func(*args, dataset=None, **kwargs,):
+#     if ray.sgd.in_worker():
+#         dataloader = dataset.get_shard(self.rank).to_torch()
+#     else:
+#         dataset = ImageDataset("code")
+#         dataloader = DataLoader(dataset)
+#     for i, (x, y) in dataloader:
+#         do_x_y(x, y)
+
+# Alternative API: SPMD call.
+# This was not chosen because it doesn't allow users to dynamically shard the dataset.
+# # Trainer will make a SPMD call.
+# model = trainer.map(
+#     training_func, 
+#     data=splits, 
+#     arg1=[X for _ in trainer.workers], 
+#     arg2=[Y for _ in trainer.workers])
+# trainer.run(validation_func, model=model, dataset=val_dataset)
+
+
+# Option 3:
+# splits = ray.dataset.filter().split(locality_hint=trainer.workers)
+# trainer.run(train_func, dataset=dataset)
+```
+</details>
+
+
+Intermediate Reporting
+----------------------
+
+Satisfies: R3 (partially), R7.
+
+Provide support for callbacks (which can be tensorboard, wandb etc). Also provides the same interface to use with Tune.
+
+```python
+def train_func(config):
+    ...
+    for _ in config["epochs"]:
+        metrics = train()
+        metrics = validate(...)
+        ray.sgd.report(**metrics)
+    return model
+
+# Higher level API
+results = trainer.run(train_func, config=config, callbacks=[...])
+
+# Lower-level generator API
+generator = trainer.run_generator(train_func, config=config)
+
+for result in generator:
+    do_stuff(result)
+
+assert generator.is_finished()
+model = generator.get_returns()
+
+
+```
+
+
+
+Hyperparameter Tuning
+---------------------
+
+Satisfies: R3.
+
+Specific requirements for hyperparameter tuning.
+
+1. Users should be able to do distributed tuning, training, and data processing.
```

Comment:
Two different scenarios to consider:

1. Tune -> Process (Dask) -> Train (SGD)
2. Process (Dask) -> Tune ->Train (SGD)

Both should be well-supported using the Ray Dataset as a bridge. The distinction being whether we want to preprocessing upfront, or we have preprocessing that varies based on hyperparam configs.